design/src/rfcs/rfc0017_logging_sensitive.md

```diff
@@ -0,0 +1,376 @@
+
+It is possible that sensitivity is a parameter passed to middleware during construction. This is similar in nature to [Use Request Extensions](#use-request-extensions) except that the `Sensitivity` is passed to middleware at startup.
+
+```rust
+struct Middleware<S> {
+    inner: S,
+    sensitivity: Sensitivity
+}
+
+impl Middleware<S> {
+    pub fn new(inner: S) -> Self { /* ... */ }
+
+    pub fn new_with_sensitivity(inner: S, sensitivity: Sensitivity) -> Self { /* ... */ }
+}
+
+impl<B, S> Service<Request<B>> for Middleware<S> {
+    /* ... */
+
+    fn call(&mut self, request: Request<B>) -> Self::Future {
+        if self.sensitivity.is_method_sensitive() {
+            debug!(method = %Sensitive(request.method()));
+        }
+
+        /* ... */
+
+        self.inner.call(request)
+    }
+}
+```
+
+It would then be required that the code generation responsible constructing a `Sensitivity` for each operation. Additionally, if any middleware is being applied to a operation then the code generation would be responsible for passing that middleware the appropriate `Sensitivity` before applying it.
+
+#### Advantages
+
+- Applicable to _all_ middleware.
+- As the `Sensitivity` struct will be known statically, the compiler will remove branches, making it potentially free.
+
+#### Disadvantages
+
+- Pollutes the API of middleware.
+
+### Redact values using a tracing Layer
+
+Distinct from `tower::Layer`, a [tracing::Layer](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/trait.Layer.html) is a "composable handler for `tracing` events". It would be possible to write an implementation which would filter out events which contain sensitive data.
+
+Examples of filtering `tracing::Layer`s already exist in the form of the [EnvFilter](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html) and [Targets](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/targets/struct.Targets.html). It is unlikely that we'll be able to leverage them for our use, but the underlying principle remains the same - the `tracing::Layer` inspects `tracing::Event`s/`tracing::Span`s filtering them based on some criteria.
+
+Code generation would be need to be used in order to produce the filtering criteria from the models. Internal developers would need to adhere to a common set of field names in order for them to be subject to the filtering. Spans would need to be opened after routing occurs in order for the `tracing::Layer` to know which operation `Event`s are being produced within and hence which filtering rules to apply.
+
+#### Advantages
+
+- Applicable to _all_ middleware.
+- Good separation of concerns:
+  - Does not pollute the API of the middleware
+  - No specific logic required within middleware.
+
+#### Disadvantages
+
+- Complex implementation.
+- Not necessarily fast.
+- `tracing::Layer`s seem to only support filtering entire `Event`s, rather than more fine grained removal of fields.
```

Comment:
Would this middleware injected by us or should the customer add it to their middleware chain?

Reply:
I think it'd have to be applied by the customer - the design of `tracing` prohibits anyone other than the customers app from constructing and initializing the tracing `Subscriber`.

Reply:
This could act as an alternative to `debug-logging` feature flag - the customer just doesn't apply the `tracing::Layer`.